Improve expereince of running linters with auto-fix capabilities on save events in editor.

### Your feature request related to a problem? Please describe

A repo I work in uses golangci-lint with the `gci` linter (among other linters).

golangci-lint is run automatically before raising each PR, and the auto-fixes from gci are applied at this point.

As a developer in this repo, I find that these changes introduce a little friction as I need to amend the suggested changes to my latest commit before raising my PR. The change which introduces the import and the change which sorts the imports may then be in different commits which introduces a little unnecessary noise in my changes (or additional effort to reorganise the commits). 

I would prefer to have the autofix applied automatically by my editor (neovim) after each save to ensure that imports are sorted as early as possible.

I can configure my editor to run gci when saving a file relatively easily. However, I would like to ensure that when gci runs on save in my editor it uses the same config that golangci-lint will use without needing to manually ensure that the two stay in sync.

I have attempted the following approaches:

1. Introduce logic in my editor configuration / an external script to parse the .golangci-lint.yml configuration and convert it into the relevant gci CLI arguments.

Limitations:
- This is quite a lot of effort for integration. For every individual linter every user will need to implement this conversion logic. This would likely be better solved once by a linter aggregator.

2. Configure my editor to run golangci-lint on save (`goclangci-lint run --fix --fast --enable-only=gci $FILENAME` flags).

Limitations:
- golangci-lint reports type check errors even in cases where gci would be able to format the file without any issues (e.g. a syntax error in another file in the package, a reference to an undefined type etc.)
- The user needs to explicitly enumerate all of the linters with autofix capabilities

### Describe the solution you'd like

1. Resolve failure modes for golangci-lint where prerequisite checks (i.e. typecheck) fail despite these presenting no issues for the enabled linters.
2. Consider adding support for only running the linters with autofix capabilities (e.g. `run --autofix-linters` flag or something similar)

### Describe alternatives you've considered

- Adding support for querying config for a specific linter (this would make the integration for users a little easier, but still non-trivial.
- Making do with running gci directly and parsing gci config manually (this does seem to have the advantage of being a bit faster)
- Simply not running the autoformatter when there are typecheck errors (loses some of the incrementality benefits of running on save.)

### Additional context

I'm happy to contribute a PR if any of the proposed solutions are acceptable

Do you have a minimal reproducible example?

Because `gci`, as a standalone app, has the same "problem" as golangci-lint with `typecheck` errors:

```
$ gci print main.go                          
Error: main.go:15:9: expected 'IDENT', found '{' (and 1 more errors)
```

For example, with the following file:
```
# gciexample.go
package gciexample

var a = b
```

Running `gci` directly:
```
$ gci print gciexample.go

package gciexample

var a = b
```

Running gci via `golangci-lint`:
```
$ golangci-lint run --fix --enable-only=gci a.go
a.go:3:9: undefined: b (typecheck)
var a = b
        ^
```

--- a/pkg/golinters/typecheck.go
+++ b/pkg/golinters/typecheck.go
@@ -20,5 +20,5 @@ func NewTypecheck() *goanalysis.Linter {
 		"Like the front-end of a Go compiler, parses and type-checks Go code",
 		[]*analysis.Analyzer{analyzer},
 		nil,
-	).WithLoadMode(goanalysis.LoadModeTypesInfo)
+	).WithLoadMode(goanalysis.LoadModeNone)
 }

--- a/pkg/lint/lintersdb/builder_linter.go
+++ b/pkg/lint/lintersdb/builder_linter.go
@@ -751,10 +751,7 @@ func (LinterBuilder) Build(cfg *config.Config) ([]*linter.Config, error) {
 		linter.NewConfig(golinters.NewTypecheck()).
 			WithInternal().
 			WithEnabledByDefault().
-			WithSince("v1.3.0").
-			WithLoadForGoAnalysis().
-			WithPresets(linter.PresetBugs).
-			WithURL(""),
+			WithSince("v1.3.0"),
 
 		linter.NewConfig(unconvert.New(&cfg.LintersSettings.Unconvert)).
 			WithSince("v1.0.0").

--- a/pkg/lint/lintersdb/manager.go
+++ b/pkg/lint/lintersdb/manager.go
@@ -204,18 +204,30 @@ func (m *Manager) build(enabledByDefaultLinters []*linter.Config) map[string]*li
 }
 
 func (m *Manager) combineGoAnalysisLinters(linters map[string]*linter.Config) {
+	mlConfig := &linter.Config{}
+
 	var goanalysisLinters []*goanalysis.Linter
 	goanalysisPresets := map[string]bool{}
+
 	for _, lc := range linters {
 		lnt, ok := lc.Linter.(*goanalysis.Linter)
 		if !ok {
 			continue
 		}
+
 		if lnt.LoadMode() == goanalysis.LoadModeWholeProgram {
 			// It's ineffective by CPU and memory to run whole-program and incremental analyzers at once.
 			continue
 		}
+
+		mlConfig.LoadMode |= lc.LoadMode
+
+		if lc.IsSlowLinter() {
+			mlConfig.ConsiderSlow()
+		}
+
 		goanalysisLinters = append(goanalysisLinters, lnt)
+
 		for _, p := range lc.InPresets {
 			goanalysisPresets[p] = true
 		}
@@ -245,22 +257,14 @@ func (m *Manager) combineGoAnalysisLinters(linters map[string]*linter.Config) {
 		return a.Name() <= b.Name()
 	})
 
-	ml := goanalysis.NewMetaLinter(goanalysisLinters)
+	mlConfig.Linter = goanalysis.NewMetaLinter(goanalysisLinters)
 
 	presets := maps.Keys(goanalysisPresets)
 	sort.Strings(presets)
+	mlConfig.InPresets = presets
 
-	mlConfig := &linter.Config{
-		Linter:           ml,
-		EnabledByDefault: false,
-		InPresets:        presets,
-		AlternativeNames: nil,
-		OriginalURL:      "",
-	}
+	linters[mlConfig.Linter.Name()] = mlConfig
 
-	mlConfig = mlConfig.WithLoadForGoAnalysis()
-
-	linters[ml.Name()] = mlConfig
 	m.debugf("Combined %d go/analysis linters into one metalinter", len(goanalysisLinters))
 }
 

--- a/pkg/lint/lintersdb/manager.go
+++ b/pkg/lint/lintersdb/manager.go
@@ -207,7 +207,6 @@ func (m *Manager) combineGoAnalysisLinters(linters map[string]*linter.Config) {
 	mlConfig := &linter.Config{}
 
 	var goanalysisLinters []*goanalysis.Linter
-	goanalysisPresets := map[string]bool{}
 
 	for _, lc := range linters {
 		lnt, ok := lc.Linter.(*goanalysis.Linter)
@@ -226,11 +225,9 @@ func (m *Manager) combineGoAnalysisLinters(linters map[string]*linter.Config) {
 			mlConfig.ConsiderSlow()
 		}
 
-		goanalysisLinters = append(goanalysisLinters, lnt)
+		mlConfig.InPresets = append(mlConfig.InPresets, lc.InPresets...)
 
-		for _, p := range lc.InPresets {
-			goanalysisPresets[p] = true
-		}
+		goanalysisLinters = append(goanalysisLinters, lnt)
 	}
 
 	if len(goanalysisLinters) <= 1 {
@@ -259,9 +256,8 @@ func (m *Manager) combineGoAnalysisLinters(linters map[string]*linter.Config) {
 
 	mlConfig.Linter = goanalysis.NewMetaLinter(goanalysisLinters)
 
-	presets := maps.Keys(goanalysisPresets)
-	sort.Strings(presets)
-	mlConfig.InPresets = presets
+	sort.Strings(mlConfig.InPresets)
+	mlConfig.InPresets = slices.Compact(mlConfig.InPresets)
 
 	linters[mlConfig.Linter.Name()] = mlConfig
 

--- a/pkg/lint/lintersdb/manager_test.go
+++ b/pkg/lint/lintersdb/manager_test.go
@@ -55,10 +55,10 @@ func TestManager_GetOptimizedLinters(t *testing.T) {
 
 	mlConfig := &linter.Config{
 		Linter:    goanalysis.NewMetaLinter(gaLinters),
-		InPresets: []string{"bugs", "format"},
+		InPresets: []string{"format"},
 	}
 
-	expected := []*linter.Config{mlConfig.WithLoadForGoAnalysis()}
+	expected := []*linter.Config{mlConfig.WithLoadFiles()}
 
 	assert.Equal(t, expected, optimizedLinters)
 }

--- a/pkg/lint/lintersdb/manager_test.go
+++ b/pkg/lint/lintersdb/manager_test.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/tools/go/packages"
 
 	"github.com/golangci/golangci-lint/pkg/config"
 	"github.com/golangci/golangci-lint/pkg/goanalysis"
@@ -176,8 +177,11 @@ func TestManager_combineGoAnalysisLinters(t *testing.T) {
 	m, err := NewManager(nil, nil)
 	require.NoError(t, err)
 
-	foo := goanalysis.NewLinter("foo", "example foo", nil, nil).WithLoadMode(goanalysis.LoadModeTypesInfo)
-	bar := goanalysis.NewLinter("bar", "example bar", nil, nil).WithLoadMode(goanalysis.LoadModeTypesInfo)
+	fooTyped := goanalysis.NewLinter("foo", "example foo", nil, nil).WithLoadMode(goanalysis.LoadModeTypesInfo)
+	barTyped := goanalysis.NewLinter("bar", "example bar", nil, nil).WithLoadMode(goanalysis.LoadModeTypesInfo)
+
+	fooSyntax := goanalysis.NewLinter("foo", "example foo", nil, nil).WithLoadMode(goanalysis.LoadModeSyntax)
+	barSyntax := goanalysis.NewLinter("bar", "example bar", nil, nil).WithLoadMode(goanalysis.LoadModeSyntax)
 
 	testCases := []struct {
 		desc     string
@@ -188,37 +192,112 @@ func TestManager_combineGoAnalysisLinters(t *testing.T) {
 			desc: "no combined, one linter",
 			linters: map[string]*linter.Config{
 				"foo": {
-					Linter:    foo,
+					Linter:    fooTyped,
 					InPresets: []string{"A"},
 				},
 			},
 			expected: map[string]*linter.Config{
 				"foo": {
-					Linter:    foo,
+					Linter:    fooTyped,
 					InPresets: []string{"A"},
 				},
 			},
 		},
 		{
-			desc: "combined, several linters",
+			desc: "combined, several linters (typed)",
 			linters: map[string]*linter.Config{
 				"foo": {
-					Linter:    foo,
+					Linter:    fooTyped,
 					InPresets: []string{"A"},
 				},
 				"bar": {
-					Linter:    bar,
+					Linter:    barTyped,
 					InPresets: []string{"B"},
 				},
 			},
 			expected: func() map[string]*linter.Config {
 				mlConfig := &linter.Config{
-					Linter:    goanalysis.NewMetaLinter([]*goanalysis.Linter{bar, foo}),
+					Linter:    goanalysis.NewMetaLinter([]*goanalysis.Linter{barTyped, fooTyped}),
 					InPresets: []string{"A", "B"},
 				}
 
 				return map[string]*linter.Config{
-					"goanalysis_metalinter": mlConfig.WithLoadForGoAnalysis(),
+					"goanalysis_metalinter": mlConfig,
+				}
+			}(),
+		},
+		{
+			desc: "combined, several linters (different LoadMode)",
+			linters: map[string]*linter.Config{
+				"foo": {
+					Linter:    fooTyped,
+					InPresets: []string{"A"},
+					LoadMode:  packages.NeedName,
+				},
+				"bar": {
+					Linter:    barTyped,
+					InPresets: []string{"B"},
+					LoadMode:  packages.NeedTypesSizes,
+				},
+			},
+			expected: func() map[string]*linter.Config {
+				mlConfig := &linter.Config{
+					Linter:    goanalysis.NewMetaLinter([]*goanalysis.Linter{barTyped, fooTyped}),
+					InPresets: []string{"A", "B"},
+					LoadMode:  packages.NeedName | packages.NeedTypesSizes,
+				}
+
+				return map[string]*linter.Config{
+					"goanalysis_metalinter": mlConfig,
+				}
+			}(),
+		},
+		{
+			desc: "combined, several linters (same LoadMode)",
+			linters: map[string]*linter.Config{
+				"foo": {
+					Linter:    fooTyped,
+					InPresets: []string{"A"},
+					LoadMode:  packages.NeedName,
+				},
+				"bar": {
+					Linter:    barTyped,
+					InPresets: []string{"B"},
+					LoadMode:  packages.NeedName,
+				},
+			},
+			expected: func() map[string]*linter.Config {
+				mlConfig := &linter.Config{
+					Linter:    goanalysis.NewMetaLinter([]*goanalysis.Linter{barTyped, fooTyped}),
+					InPresets: []string{"A", "B"},
+					LoadMode:  packages.NeedName,
+				}
+
+				return map[string]*linter.Config{
+					"goanalysis_metalinter": mlConfig,
+				}
+			}(),
+		},
+		{
+			desc: "combined, several linters (syntax)",
+			linters: map[string]*linter.Config{
+				"foo": {
+					Linter:    fooSyntax,
+					InPresets: []string{"A"},
+				},
+				"bar": {
+					Linter:    barSyntax,
+					InPresets: []string{"B"},
+				},
+			},
+			expected: func() map[string]*linter.Config {
+				mlConfig := &linter.Config{
+					Linter:    goanalysis.NewMetaLinter([]*goanalysis.Linter{barSyntax, fooSyntax}),
+					InPresets: []string{"A", "B"},
+				}
+
+				return map[string]*linter.Config{
+					"goanalysis_metalinter": mlConfig,
 				}
 			}(),
 		},

--- a/test/enabled_linters_test.go
+++ b/test/enabled_linters_test.go
@@ -148,7 +148,7 @@ func getEnabledByDefaultFastLintersExcept(t *testing.T, except ...string) []stri
 	ebdl := m.GetAllEnabledByDefaultLinters()
 	var ret []string
 	for _, lc := range ebdl {
-		if lc.IsSlowLinter() {
+		if lc.IsSlowLinter() || lc.Internal {
 			continue
 		}
 
@@ -169,7 +169,7 @@ func getAllFastLintersWith(t *testing.T, with ...string) []string {
 	linters := dbManager.GetAllSupportedLinterConfigs()
 	ret := append([]string{}, with...)
 	for _, lc := range linters {
-		if lc.IsSlowLinter() {
+		if lc.IsSlowLinter() || lc.Internal {
 			continue
 		}
 		ret = append(ret, lc.Name())
@@ -206,7 +206,7 @@ func getEnabledByDefaultFastLintersWith(t *testing.T, with ...string) []string {
 	ebdl := dbManager.GetAllEnabledByDefaultLinters()
 	ret := append([]string{}, with...)
 	for _, lc := range ebdl {
-		if lc.IsSlowLinter() {
+		if lc.IsSlowLinter() || lc.Internal {
 			continue
 		}
 

